asyncssh/stream.py

```diff
@@ -602,7 +613,7 @@ async def readuntil(self, separator: object, datatype: DataType) -> AnyStr:
 
                     newbuf = cast(AnyStr, recv_buf[curbuf])
                     buf += newbuf
-                    start = max(buflen + 1 - seplen, 0)
+                    start = 0 if seplen is None else max(buflen + 1 - seplen, 0)
```

Comment:
mmmm... Not sure I follow you.
How can I set the length on my own?
If the user didn't provide the max-len, I have 2 options:
1. Validate it and fail when not provided
2. try to match on the whole buffer.

The `start` variable is dependent on the `buflen`, which is not a constant.

Maybe I'm missing something.
Can you please elaborate?

Reply:
I was suggesting that it should fail with something like a ValueError exception if the caller passed in a regex as `separator` but failed to also pass in a `max_separator_len`. If the check for that was done in earlier code, the `seplen` value here should never be `None` and the existing code at this line should be sufficient.

Reply:
Actually, after re-reading the code tonight, I take back what I said about there being a correctness issue in the case where no max separator length is specified. As you said, setting that provides a performance benefit, but it will still find matches across buffer boundaries if it's not set with the code you have above. i was thinking it would end up only matching within each new buffer in that case, not holding onto enough data from earlier buffers to properly match across the boundary. Given that, I take back my suggestion here to require max_separator_len to be specified.

Reply:
👍 

If you want me to add a warning in the docstring let me know.

Reply:
Thanks - there's some other minor cleanup I'll probably do after merging this. I can take care of it then. There are really two issues I'm thinking of calling out here. One is around the cost of not limiting the regex match to a small size by specifying the max_separator_len, and the other my point above about some regex patterns not always matching the longest possible string if the match is broken across a buffer boundary.